// node-scratch-client/src/Struct/Studio.js
const Image = require("./Image.js");
const request = require("../request.js");

class Studio {
  constructor(Client, raw) {
    this._client = Client;

    this.id = raw.id;
    this.title = raw.title;
    this.owner = raw.owner;
    this.description = raw.description;
    this.image = new Image(raw.image);

    this.createdTimestamp = raw.history.created;
    this.lsatModifiedTimestamp = raw.history.modified;

    this.followerCount = raw.stats.followers;
  }

  addProject(project) {
    let id = project.id || project;

    let _this = this;

    return new Promise((resolve, reject) => {
      request({
        path: "/site-api/projects/in/" + _this.id + "/add/?pks=" + id,
        method: "PUT",
        sessionid: _this._client.session.sessionid,
        csrftoken: _this._client.session.csrftoken
      }, {
        accept: "application/json",
        "Content-Type": "application/json",
        origin: "https://scratch.mit.edu",
        referer: "https://scratch.mit.edu/studios/" + _this.id + "/",
        "X-Token": _this._client.session.authorized.user.accessToken
      }).then(response => {
        resolve();
      }).catch(reject);
    });
  }
}

module.exports = Studio;
